File: leetcode/图/其他/_1615_最大网络秩.cpp
#include<vector>
using namespace std;

// https://leetcode.cn/problems/maximal-network-rank/
class Solution {
public:
    int maximalNetworkRank(int n, vector<vector<int>>& roads) {
        vector<int> degrees(n);
        vector<vector<int>> graph(n,vector<int>(n));
        for(auto road : roads){
            degrees[road[0]]++;
            degrees[road[1]]++;
            graph[road[0]][road[1]] = 1;
            graph[road[1]][road[0]] = 1;
        }
        int maxRank = 0;
        for(int i = 0; i < n; i++){
            for(int j = i + 1; j < n; j++){
                maxRank = max(maxRank,degrees[i] + degrees[j] - graph[i][j]);
            }
        }
        return maxRank;
    }
};

File: leetcode/_每日一题/_1615_最大网络秩.py
# https://leetcode.cn/problems/maximal-network-rank/
class Solution:
    def maximalNetworkRank(self, n: int, roads: list) -> int:
        d = [0 for _ in range(n)]
        g = [[0 for _ in range(n)] for _ in range(n)]
        for i, j in roads:
            g[i][j] = g[j][i] = 1
            d[i], d[j] = d[i] + 1, d[j] + 1
        maxRank = 0
        for i in range(n):
            for j in range(i + 1, n):
                maxRank = max(maxRank, d[i] + d[j] - g[i][j])
        return maxRank